README.md
# zoomer
[![pub package](https://img.shields.io/pub/v/zoomer.svg)](https://pub.dartlang.org/packages/zoomer)

Zoomer - To Zoom Any Widget

## preview
<img src="https://raw.githubusercontent.com/RutvikRana/zoomer/main/video_example/96.gif" alt="" width="200" height="200">

## What It Can Do?
* Zoom
* Rotate
* Translate

## Installation
Follow Installation guide of Pub.dev

## Syntax

1. Zoomer Class

        Zoomer(
            {this.child,                    // Child Can Be Any Widget, you want to make zoomable
            this.controller,                // ZoomerControoller To Control Zoomer
            this.height,                    // Height
            this.width,                     // Width 
            this.background,                // background of Zoomer. It is BoxDecoration.
            this.maxScale = 2.0,            // maximum Scale
            this.minScale = 0.5,            // mininum Scale
            this.enableTranslation = false, // Want to move it by touch
            this.enableRotation = false,    // Want to rotate it
            this.clipRotation = true});     // Clips the Rotation to 90-degrees

2. ZoomerController Class

          ZoomerController({initialScale = 1.0})  //To Set initial Scale
          
    * APIs
    
      - double get scale              
      ```scale = _controller.scale```
      
      - set setScale(double value)    
       ```_controller.setScale = 1.5```
       
      - Offset get offset             
      ```offset = _controller.offset```
      
      - set setOffset(Offset value)   
       ```_controller.setOffset = Offset(0,10)```
       
      - double get rotation           
      ```rotation = _controller.rotation```
      
      - set setRotation(double value)               
      ```_controller.setRotation = pi/4```
      **Note: Rotation is in Radians**    
      
      - onZoomStart()             
      ```_controller.onZoomStart( (){ print( _controller.scale ); } )```

      - onZoomUpdate()             
      ```_controller.onZoomUpdate( (){ print( _controller.scale ); } )```

      - onZoomEnd()             
      ```_controller.onZoomEnd( (){ print( _controller.scale ); } )```

## Example

```class _HomeState extends State<Home> {
  ZoomerController _zoomerController = ZoomerController(initialScale: 1.0);
  String _zoomDetails = "Zoom";  

  @override
  Widget build(BuildContext context) {

    _zoomerController.onZoomUpdate((){
      setState(() {
        _zoomDetails = "Scale = "+ _zoomerController.scale.toStringAsFixed(2);
        _zoomDetails += "\nRotation = "+ _zoomerController.rotation.toStringAsFixed(2);
        _zoomDetails += "\nOffset = ("+ _zoomerController.offset.dx.toStringAsFixed(2)+","+_zoomerController.offset.dy.toStringAsFixed(2)+")";
      });
    });

    return Scaffold(
      appBar: AppBar(title: Text("Zommer Example"),),
      body:
        Center(child:
        Stack(
          children: [
            Align(alignment: Alignment.topCenter,child: SizedBox(height: 150,child: Text(_zoomDetails,textAlign: TextAlign.center,style: TextStyle(fontSize: 30),))),
            Center(
              child: 
                Zoomer(
                  enableTranslation: true,
                  enableRotation: true,
                  clipRotation: true,
                  maxScale: 2,
                  minScale: 0.5,
                  background:BoxDecoration(color: Colors.white),
                  height: 300,
                  width: 300,
                  controller: _zoomerController,
                  child: Container(decoration: BoxDecoration(color: Colors.green),height: 200,width: 200,child: FlutterLogo(),)),
            ),
          ])),
    );
  }
}
```

## Contact Me

I Am Rutvik Rana, Medical Student cum Passionate Coder, Invite You To My [Coding(noob to pro)](https://t.me/coding_noob_to_pro) Channel.


CHANGELOG.md
## [0.0.4]

* Fixed ZoomerController Class

## [0.0.3]

* Fixed Example Gif for pub readme.md

## [0.0.2]

* Improved Documentation.

## [0.0.1]

* Zooomer Landed on Pub.dev.


test/zoomer_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:zoomer/zoomer.dart';

main()=>runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Zoomer Test",
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Zoomer Test"),),
      body: SafeArea(child:
        Container(child:
          Zoomer(child: Container(height: 200,width: 200,color: Colors.blueAccent,child: CircleAvatar(),),),),),
    );
  }
}


lib/zoomer.dart
import 'dart:math';

import 'package:flutter/material.dart';

class ZoomerController{
  double _initScale;
  void Function(double value) _setZoomerScale,_setZoomerAngle;
  double Function() _getZoomerScale,_getZoomerAngle;
  void Function(Offset value) _setZoomerOffset;
  Offset Function() _getZoomerOffset;
  Function _onZoomStart,_onZoomUpdate,_onZoomEnd;
  
  ///ZoomerController To Control Zoomer Widget
  ZoomerController({initialScale = 1.0}){
    _initScale = initialScale;
  }

  void onZoomStart(Function zoom){
    _onZoomStart = zoom;
  }
  void onZoomUpdate(Function zoom){
    _onZoomUpdate = zoom;
  }
  void onZoomEnd(Function zoom){
    _onZoomEnd = zoom;
  }
  
  double get scale{if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");} return _getZoomerScale();}
  set setScale(double value){if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");}_setZoomerScale(value);}
  Offset get offset{if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");}return _getZoomerOffset();}
  set setOffset(Offset value){if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");}_setZoomerOffset(value);}
  double get rotation{if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");}return _getZoomerAngle();}
  set setRotation(double value){if(_getZoomerScale==null){throw Exception("ZoomerController Not Attached To Any Zoomer");}_setZoomerAngle(value);}

}

class Zoomer extends StatefulWidget {
  ///Zoomer widget to create interactive Zoomable widget
  final Widget child;
  final double height,width,minScale,maxScale;
  final BoxDecoration background;
  final bool enableRotation;
  final bool clipRotation;
  final bool enableTranslation;
  final ZoomerController controller;
  Zoomer({this.child,this.controller,this.height,this.width,this.background,this.maxScale = 2.0,this.minScale = 0.5,this.enableTranslation = false,this.enableRotation = false,this.clipRotation = true});
  @override
  _ZoomerState createState() => _ZoomerState();
}

class _ZoomerState extends State<Zoomer> {
  double _scale = -1.0;
  Offset _offset = Offset(0.0, 0.0);
  double _startScale = 1.0;
  Offset _lastOffset;
  Offset _startOffset;
  Offset _limitOffset;
  double _angle = 0.0;
  double _startAngle = 0.0;

  double get getScale=>-_scale;
  set setScale(double value)=> setState((){_scale = -value;});
  Offset get getOffset=>_offset;
  set setOffset(Offset value)=>setState((){_offset = value;});
  double get getRotation=>_angle;
  set setRotation(double value)=>setState((){_angle = value;});

  @override
  void initState(){
    super.initState();
    double l = (1+_scale);
     l = l<0?(-_scale-1)/4:l;
    _limitOffset = Offset(widget.width, widget.height)*l;
    if(widget.controller != null){
      _scale = -widget.controller._initScale;
      widget.controller._getZoomerScale = ()=>getScale;
      widget.controller._getZoomerAngle = ()=>getRotation;
      widget.controller._getZoomerOffset = ()=>getOffset;
      widget.controller._setZoomerScale = (value){setScale = value;};
      widget.controller._setZoomerAngle = (value){setRotation = value;};
      widget.controller._setZoomerOffset = (value){setOffset = value;};

    }
  }

  void _scaleStart(ScaleStartDetails details){

    if(widget.controller != null){
      if(widget.controller._onZoomStart != null){
        widget.controller._onZoomStart();
    }}

    _lastOffset = details.focalPoint;
    _startScale = -_scale;
    _startOffset = _offset;
    _startAngle = _angle;
  }  

  void _scaleEnd(ScaleEndDetails details){

    if(widget.enableTranslation){
      double l = (1+_scale);
      l = l<0?(-_scale-1)/4:l;
      _limitOffset = Offset(widget.width, widget.height)*l;
    }

    if(widget.enableRotation){
      if(widget.clipRotation){
        double x = ((_angle % (2*pi))/(pi/2));
        double angle = ((x-x.floor())>0.5?(x.floor()+1):x.floor()) * (pi/2);
        setState(() {
          _angle= angle;
        });
      }
    }

    if(widget.controller != null){
      if(widget.controller._onZoomEnd != null){
        widget.controller._onZoomEnd();
    }}
  }

  void _scaleUpdate(ScaleUpdateDetails details){
    double scale = (_startScale * details.scale).clamp(widget.minScale,widget.maxScale);    
    Offset offset = details.focalPoint - ((_lastOffset - _startOffset)/_startScale) * scale;
    double angle = _startAngle+details.rotation;
    setState(() {
      _scale = -scale;

      if(widget.enableRotation){
      _angle = angle;
      }

      if(details.scale == 1.0 && widget.enableTranslation){
        _offset = Offset(offset.dx.clamp(-_limitOffset.dx, _limitOffset.dx),offset.dy.clamp(-_limitOffset.dy, _limitOffset.dy));
      }
    });

    if(widget.controller != null){
      if(widget.controller._onZoomUpdate != null){
        widget.controller._onZoomUpdate();
    }}
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onScaleStart: _scaleStart,
      onScaleUpdate: _scaleUpdate,
      onScaleEnd: _scaleEnd,
      child: ClipRect(
        child: Container(
          decoration: widget.background==null?BoxDecoration(color: Colors.black):widget.background,
          height: widget.height,
          width: widget.width,
            child: Transform(transform:Matrix4.identity()..scale(-_scale,-_scale)..translate(_offset.dx,_offset.dy)..rotateZ(_angle),
              alignment: FractionalOffset.center,
              child:widget.child,
            ),
          ),
      ),
    );
  }
}


Example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'zoomer.dart';

void main() => runApp(MyApp());

// ------ Root Widget ---------
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Zoomer Example",
      theme: ThemeData(
          primarySwatch: Colors.green,
          canvasColor: Colors.blue.shade100,
          visualDensity: VisualDensity.adaptivePlatformDensity,
          platform: TargetPlatform.android),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  ZoomerController _zoomerController = ZoomerController(initialScale: 1.0);
  String _zoomDetails = "Zoom";  

  @override
  Widget build(BuildContext context) {

    _zoomerController.onZoomUpdate((){
      setState(() {
        _zoomDetails = "Scale = "+ _zoomerController.scale.toStringAsFixed(2);
        _zoomDetails += "\nRotation = "+ _zoomerController.rotation.toStringAsFixed(2);
        _zoomDetails += "\nOffset = ("+ _zoomerController.offset.dx.toStringAsFixed(2)+","+_zoomerController.offset.dy.toStringAsFixed(2)+")";
      });
    });

    return Scaffold(
      appBar: AppBar(title: Text("Zommer Example"),),
      body:
        Center(child:
        Stack(
          children: [
            Align(alignment: Alignment.topCenter,child: SizedBox(height: 150,child: Text(_zoomDetails,textAlign: TextAlign.center,style: TextStyle(fontSize: 30),))),
            Center(
              child: Zoomer(
                enableTranslation: true,
                enableRotation: true,
                clipRotation: true,
                maxScale: 2,
                minScale: 0.5,
                background:BoxDecoration(color: Colors.white),
                height: 300,
                width: 300,
                controller: _zoomerController,
                child: Container(decoration: BoxDecoration(color: Colors.green),height: 200,width: 200,child: FlutterLogo(),)),
            ),
          ])),
    );
  }
}
